trulens_eval/trulens_eval/feedback/provider/endpoint/cortex.py

```diff
@@ -0,0 +1,92 @@
+import inspect
+import logging
+import pprint
+from typing import Any, Callable, ClassVar, Optional
+import json
+
+
+from trulens_eval.feedback.provider.endpoint.base import Endpoint
+from trulens_eval.feedback.provider.endpoint.base import EndpointCallback
+from trulens_eval.utils.imports import OptionalImports
+from trulens_eval.utils.imports import REQUIREMENT_CORTEX
+
+logger = logging.getLogger(__name__)
+
+pp = pprint.PrettyPrinter()
+
+with OptionalImports(messages=REQUIREMENT_CORTEX):
+    from snowflake.snowpark import Session, DataFrame
+
+
+class CortexCallback(EndpointCallback):
+    model_config: ClassVar[dict] = dict(arbitrary_types_allowed=True)
+
+    # TODO: confirm if we really don't need to implement handle_classification
+
+    def handle_generation(self, response: dict) -> None:
+        """Get the usage information from Cortex LLM function response's usage field."""
+        usage = response['usage']
+
+        # Incremente number of requests.
+        super().handle_generation(response)
+
+        # Assume a response that had usage field was successful. Note at the time of writing 06/12/2024, the usage
+        # information from Cortex LLM functions is only available when called via snow SQL. It's not fully supported in
+        # Python API such as `from snowflake.cortex import Summarize, Complete, ExtractAnswer, Sentiment, Translate` not yet.
+
+        self.cost.n_successful_requests += 1
+
+        for cost_field, cortex_field in [
+            ("n_tokens", "total_tokens"),
+            ("n_prompt_tokens", "prompt_tokens"),
+            ("n_completion_tokens", "completion_tokens"),
+        ]:
+            setattr(self.cost, cost_field, usage.get(cortex_field, 0))
```

Comment:
Can you explain this to me? The `openai.py` does the same thing, but shouldn't this add this to the `cost_field`? If we know it's the only request, why would we ever increment `n_requests` in L31 or `n_successful_requests` in L37?

Reply:
re: "If we know it's the only request, why would we ever increment n_requests in L31 or n_successful_requests in L37?"  My understanding is that even if this is the case for the current Cortex LLM function, it might change in the future if Cortex is also trying to introduce some sort of agentic abstraction layer. 

One example I can think of is say trulens users are using LlamaIndex's `OpenAIAgent` wrapper, and there could be multiple generation (completion) calls per wrapped / instrumented call in TruLens, if the trulens users decided to instrument the tool calling components.

Reply:
For the cost tracking - checked with Piotr as well and I believe we should fix it